x/axelarnet/message_handler.go

```diff
@@ -21,17 +22,38 @@ import (
 
 // Fee is used to pay relayer for executing cross chain message
 type Fee struct {
-	Amount    string `json:"amount"`
-	Recipient string `json:"recipient"`
+	Amount          string  `json:"amount"`
+	Recipient       string  `json:"recipient"`
+	RefundRecipient *string `json:"refund_recipient"`
 }
 
-func (f Fee) validate(ctx sdk.Context, n types.Nexus, b types.BankKeeper, token sdk.Coin, msgType nexus.MessageType) error {
-	amt, ok := sdk.NewIntFromString(f.Amount)
-	if !ok || amt.LTE(sdk.ZeroInt()) {
+func (f Fee) validateBasic() error {
+	amount, ok := sdk.NewIntFromString(f.Amount)
+	if !ok || !amount.IsPositive() {
 		return fmt.Errorf("invalid fee amount")
 	}
 
-	afterFee := token.Amount.Sub(amt)
+	if _, err := sdk.AccAddressFromBech32(f.Recipient); err != nil {
+		return err
+	}
+
+	if f.RefundRecipient != nil {
+		// The refund recipient is an address of the chain where the request is originally initiated from,
+		// and therefore we cannot validate much of it.
+		if err := utils.ValidateString(*f.RefundRecipient); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f Fee) validate(ctx sdk.Context, n types.Nexus, b types.BankKeeper, token sdk.Coin, msgType nexus.MessageType) error {
```

Comment:
this is a strange mashup function. From what I can tell it does:
1. check the fee is in a valid format (this is the wanted fee, not the paid fee, right?)
2. check that the fee is covered
3. check that the token denomination is registered
4. check the fee recipient is not blocked

some of these things are already checked in `validateMessage` I believe. It the very least, it doesn't feel like a `Fee` method but a `validateFee(ctx, fee, ...)` function, or maybe a `feeValidator.IsAllowed(...)`. However, I'd prefer if we could break it up into it's components

Reply:
I'll do the `validateFee(...)`. Didn't put much thought into this since I was only trying to extract the basic validation out, but you made a good point that this doesn't make sense to me Fee method.